### PR TITLE
Production code changes to aid with test stabilization

### PR DESCRIPTION
**Pull request by sharwell**
This pull request contains the changes to production code resulting from a larger effort to stabilize flaky tests.

* Move the listener/waiter test hooks to the workspaces layer
* Cancel pending asynchronous operations when `IdleProcessor` stops
* Extract interface `ISerializer` from `Serializer`
* Convert `ISerializer` to workspace service `ISerializerService`
* Add service creation hooks for `HostServices` implementations
* Avoid possible MEF composition in `Workspace.Dispose`
* Make `AssetService.s_serializer` a per-instance field
* Make `PrimaryWorkspace` a MEF component instead of static
* Make `SolutionService.PrimaryWorkspace` a per-instance property
* Add the ability to hook `RoslynServices.HostServices` during testing
* Place `ITextBufferCloneService` in `ITextBuffer.Properties` instead of separate tracking in `SnapshotSourceText`

### Customer scenario

This is a change to improve the testability of our code. It is not intended to produce any user-visible behavior or performance changes.

### Bugs this fixes

n/a (but supports fixing many flaky test issues)

### Workarounds, if any

None

### Risk

Relatively low. The risk here primarily lies in a few changes to internal APIs. Downstream consumers accessing `PrimaryWorkspace`, `Serializer`, or a couple Roslyn remoting types through IVT will need to update their code to avoid this path.

### Performance impact

No expected performance impact in production.

### Is this a regression from a previous update?

This is a developer/contributor regression. Our tests have become more sensitive to mutable data sharing across test boundaries as we adopt updated dependencies that impose stricter requirements for asynchronous code, increasing the chances that tests will fail in CI builds.

### Root cause analysis

See previous.

### How was the bug found?

CI test process dump analysis.

### Test documentation updated?

No change to testing approaches results directly from this change.

### CODE REVIEW

File: src/Features/Core/Portable/Shared/TestHooks/Legacy/ListenerForwarders.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+using System.Runtime.CompilerServices;
+
+[assembly: TypeForwardedTo(typeof(Microsoft.CodeAnalysis.Shared.TestHooks.AsynchronousOperationListener))]
+[assembly: TypeForwardedTo(typeof(Microsoft.CodeAnalysis.Shared.TestHooks.AsynchronousOperationListenerProvider))]
+[assembly: TypeForwardedTo(typeof(Microsoft.CodeAnalysis.Shared.TestHooks.EmptyAsyncToken))]
+[assembly: TypeForwardedTo(typeof(Microsoft.CodeAnalysis.Shared.TestHooks.IAsynchronousOperationListenerProvider))]
+[assembly: TypeForwardedTo(typeof(Microsoft.CodeAnalysis.Shared.TestHooks.IAsynchronousOperationWaiter))]
+[assembly: TypeForwardedTo(typeof(Microsoft.CodeAnalysis.Shared.TestHooks.IAsynchronousOperationListener))]
+[assembly: TypeForwardedTo(typeof(Microsoft.CodeAnalysis.Shared.TestHooks.IAsyncToken))]
+[assembly: TypeForwardedTo(typeof(Microsoft.CodeAnalysis.Shared.TestHooks.TaskExtensions))]
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Shared/TestHooks/Legacy/ListenerForwarders.cs]: Out of curiosity, why are these needed?  for F#/TS/JS?

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs [hunk 2]
```diff
@@ -48,6 +48,20 @@ private sealed class SemanticChangeProcessor : IdleProcessor
...
+                            foreach (var (documentId, data) in _pendingWork)
+                            {
+                                data.AsyncToken.Dispose();
+                            }
+
+                            _pendingWork.Clear();
+                        },
+                        CancellationToken.None,
+                        TaskContinuationOptions.ExecuteSynchronously,
+                        TaskScheduler.Default);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: Would benefit from comments explaining what's going on here.  A comment also explaining that this should run unilaterally, regardless of how AsyncProcessorTask completes might be good.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs [hunk 3]
```diff
@@ -339,6 +353,20 @@ private class ProjectProcessor : IdleProcessor
...
+                                foreach (var (projectId, data) in _pendingWork)
+                                {
+                                    data.AsyncToken.Dispose();
+                                }
+
+                                _pendingWork.Clear();
+                            },
+                            CancellationToken.None,
+                            TaskContinuationOptions.ExecuteSynchronously,
+                            TaskScheduler.Default);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: duplicate code.  Can we we share with above?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: Should we be using LazyCancellation here?  in other words, is it ok if this code runs before some antecedent tasks completes, but after the cancellation token controlling another intermediary tasks completes?  In other words, should this be using the semantics of SafeContinueWith?

File: src/Workspaces/Core/Portable/Execution/SerializerService.cs [hunk 4]
```diff
@@ -29,15 +30,8 @@ internal partial class Serializer
...
-        {
-        }
-
-        public Serializer(Workspace workspace) : this(workspace.Services)
-        {
-        }
-
-        public Serializer(HostWorkspaceServices workspaceServices)
+        [Obsolete("This exported object must be obtained through the MEF export provider.", error: true)]
+        public SerializerService(HostWorkspaceServices workspaceServices)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/SerializerService.cs]: cute.  i like that.  we should do that in a lot of places :)

File: src/Workspaces/Core/Portable/Execution/SerializerServiceFactory.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Serialization;
+
+namespace Microsoft.CodeAnalysis.Execution
+{
+    [ExportWorkspaceServiceFactory(typeof(ISerializerService), layer: ServiceLayer.Default), Shared]
+    internal class SerializerServiceFactory : IWorkspaceServiceFactory
+    {
+        [Obsolete("This is the factory method for " + nameof(SerializerService) + ".", error: true)]
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/SerializerServiceFactory.cs]: Could we have our own subclass for this attribute?  it reads as if: "this is obsolete and hsould not be used.  people should transition to something else".  But, in reality, it just means:  "it's only for use by mef.  we want to keep it, but you should not reference it directly in your code."

Something like "MefConstructorAttribute : ObsoleteAttribute"?

Comment (sharwell) [reply src/Workspaces/Core/Portable/Execution/SerializerServiceFactory.cs]: ➡️ `ObsoleteAttribute` is sealed, but I'm considering creating a class that holds a `const string` for the message with a comment on that declaration.

File: src/Features/Core/Portable/Shared/TestHooks/Legacy/ListenerForwarders.cs (hunk 1 above)
Comment (sharwell) [reply src/Features/Core/Portable/Shared/TestHooks/Legacy/ListenerForwarders.cs]: ➡️ Considering the impact of other changes coming in this PR, I'm guessing these are not going to be necessary in the final revision. However, I decided to make the relocation separate from the removal of the original types.

File: src/Workspaces/Core/Portable/Execution/SerializerService.cs (hunk 4 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Execution/SerializerService.cs]: ➡️ #25560

File: src/Workspaces/Core/Portable/Workspace/PrimaryWorkspace.cs [hunk 6]
```diff
@@ -1,22 +1,27 @@
...
 
 namespace Microsoft.CodeAnalysis
 {
-    internal static class PrimaryWorkspace
+    [Export(typeof(PrimaryWorkspace)), Shared]
+    internal sealed class PrimaryWorkspace
     {
-        private static readonly ReaderWriterLockSlim s_registryGate = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
-        private static Workspace s_primaryWorkspace;
+        private readonly ReaderWriterLockSlim s_registryGate = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
```
Comment (jaredpar) [inline src/Workspaces/Core/Portable/Workspace/PrimaryWorkspace.cs]: >s_registryGate [](start = 46, length = 14)

s_ prefix no longer needed as this is an instance variable.

File: src/Workspaces/Core/Portable/Workspace/PrimaryWorkspace.cs [hunk 7]
```diff
@@ -1,22 +1,27 @@
...
+        private Workspace s_primaryWorkspace;
```
Comment (jaredpar) [inline src/Workspaces/Core/Portable/Workspace/PrimaryWorkspace.cs]: >s_primaryWorkspace [](start = 26, length = 18)

s_ prefix no longer needed as this is an instance variable.

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.Connections.cs [hunk 8]
```diff
@@ -154,16 +157,16 @@ private static void ReportTimeout(Stopwatch watch)
 
             private static bool s_infoBarReported = false;
 
-            private static void ShowInfoBar()
+            private static void ShowInfoBar(Workspace workspace)
```
Comment (jaredpar) [inline src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.Connections.cs]: Shouldn't we assert when this is null? The API is a no-op if this is null hence seems like a required paramter.

File: src/Workspaces/Core/Desktop/Workspace/Host/Mef/MefV1HostServices.cs [hunk 9]
```diff
@@ -39,18 +43,28 @@ public static MefV1HostServices Create(ExportProvider exportProvider)
...
+            {
+                return s_CreationHook(assemblies);
+            }
+
             var catalog = new AggregateCatalog(assemblies.Select(a => new AssemblyCatalog(a)));
             var container = new CompositionContainer(catalog, compositionOptions: CompositionOptions.DisableSilentRejection | CompositionOptions.IsThreadSafe);
             return new MefV1HostServices(container);
         }
 
+        internal static void HookServiceCreation(CreationHook hook)
```
Comment (jaredpar) [inline src/Workspaces/Core/Desktop/Workspace/Host/Mef/MefV1HostServices.cs]: > internal static void HookServiceCreation(CreationHook hook) [](start = 7, length = 60)

Should we assert that `s_CreationHook` is `null`? Lacking that seems like we will risk introducing bugs by silently overwriting the previous vaule.

Comment (sharwell) [reply src/Workspaces/Core/Desktop/Workspace/Host/Mef/MefV1HostServices.cs]: ➡️ No. In the cleanup step for the test hook, the creation hook is replaced by a non-null hook that produces an informative exception/test failure in the event code attempts to create an ExportProvider after the test exits the isolation boundary.

**Comment by sharwell** [comment]
:memo: @CyrusNajmabadi @jaredpar I'm going to apply code changes resulting from code review after I work with HeeJae to complete the final behavior step.

File: src/Workspaces/Core/Desktop/Workspace/Host/Mef/MefV1HostServices.cs (hunk 9 above)
Comment (jaredpar) [reply src/Workspaces/Core/Desktop/Workspace/Host/Mef/MefV1HostServices.cs]: For cases where the overwrite is inteended a `force` parameter could be employed. 


---
In reply to: [175523407](https://github.com/dotnet/roslyn/pull/25558#discussion_r175523407) [](ancestors = 175523407)

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs [hunk 10]
```diff
@@ -48,6 +48,20 @@ private sealed class SemanticChangeProcessor : IdleProcessor
                     _pendingWork = new Dictionary<DocumentId, Data>();
 
                     Start();
+
+                    AsyncProcessorTask.ContinueWith(
+                        _ =>
+                        {
+                            foreach (var (documentId, data) in _pendingWork)
```
Comment (jaredpar) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: >documentId, [](start = 42, length = 11)

Consider using `_` instead of `documentId` to make it clear it's not used.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs [hunk 11]
```diff
@@ -339,6 +353,20 @@ private class ProjectProcessor : IdleProcessor
                         _pendingWork = new Dictionary<ProjectId, Data>();
 
                         Start();
+
+                        AsyncProcessorTask.ContinueWith(
+                            _ =>
+                            {
+                                foreach (var (projectId, data) in _pendingWork)
```
Comment (jaredpar) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: >projectId [](start = 46, length = 9)

consider using `_` instead of `projecTId` to make it clear it's not used.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs [hunk 12]
```diff
@@ -48,6 +48,20 @@ private sealed class SemanticChangeProcessor : IdleProcessor
...
+
+                    AsyncProcessorTask.ContinueWith(
+                        _ =>
+                        {
+                            foreach (var (documentId, data) in _pendingWork)
+                            {
+                                data.AsyncToken.Dispose();
+                            }
+
+                            _pendingWork.Clear();
```
Comment (jaredpar) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: >_pendingWork.Clear(); [](start = 28, length = 21)

Why aren't the operations on `_pendingWork` guarded by the appropriate gate here?

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphProvider.cs [hunk 13]
```diff
@@ -19,19 +19,21 @@ internal class AbstractGraphProvider : IGraphProvider
...
         private readonly IServiceProvider _serviceProvider;
+        private readonly Workspace _workspace;
         private readonly GraphQueryManager _graphQueryManager;
 
         private bool _initialized = false;
 
         protected AbstractGraphProvider(
             IGlyphService glyphService,
             SVsServiceProvider serviceProvider,
-            CodeAnalysis.Workspace workspace,
```
Comment (jaredpar) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphProvider.cs]: > CodeAnalysis.Workspace workspace, [](start = 11, length = 34)

A bit disturbing this parameter was provided before and simply ignored.

File: src/Workspaces/Core/Portable/Workspace/PrimaryWorkspace.cs [hunk 14]
```diff
@@ -1,22 +1,27 @@
...
 using System;
-using System.Collections.Generic;
+using System.Composition;
 using System.Threading;
-using System.Threading.Tasks;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis
 {
-    internal static class PrimaryWorkspace
```
Comment (jaredpar) [inline src/Workspaces/Core/Portable/Workspace/PrimaryWorkspace.cs]: >internal static class PrimaryWorkspace [](start = 4, length = 38)

I support all removal of global mutable statics.

File: src/Workspaces/Core/Desktop/Workspace/Host/Mef/MefV1HostServices.cs [hunk 15]
```diff
@@ -17,6 +17,10 @@ namespace Microsoft.CodeAnalysis.Host.Mef
     /// </summary>
     public class MefV1HostServices : HostServices, IMefHostExportProvider
     {
+        internal delegate MefV1HostServices CreationHook(IEnumerable<Assembly> assemblies);
+
+        private static CreationHook s_CreationHook;
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Desktop/Workspace/Host/Mef/MefV1HostServices.cs]: Would like a comment that this is test-only. I can imagine people doing really terrible things with this thinking this is a "feature".

File: src/Workspaces/Core/Portable/Workspace/Host/Mef/MefHostServices.cs [hunk 16]
```diff
@@ -67,6 +79,12 @@ private class WithMetadataImporter<TExtension, TMetadata>
 
         #region Defaults
 
+        internal static void HookServiceCreation(CreationHook hook)
+        {
+            s_CreationHook = hook;
+            s_defaultHost = null;
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Host/Mef/MefHostServices.cs]: Comment why. I can guess, but I'm not necessarily the next person to touch this. :smile:

File: src/Workspaces/Core/Portable/Workspace/Host/Mef/MefHostServices.cs [hunk 17]
```diff
@@ -13,6 +13,10 @@ namespace Microsoft.CodeAnalysis.Host.Mef
 {
     public class MefHostServices : HostServices, IMefHostExportProvider
     {
+        internal delegate MefHostServices CreationHook(IEnumerable<Assembly> assemblies, bool requestingDefaultHost);
+
+        private static CreationHook s_CreationHook;
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Host/Mef/MefHostServices.cs]: Ditto: comment this is test only. Even better: name the field appropriately.

File: src/EditorFeatures/Text/Extensions.cs [hunk 18]
```diff
@@ -12,7 +12,7 @@ namespace Microsoft.CodeAnalysis.Text
     public static partial class Extensions
     {
         public static SourceTextContainer AsTextContainer(this ITextBuffer buffer)
-            => TextBufferContainer.From(buffer);
+            => TextBufferContainer.From(workspace: null, buffer);
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Text/Extensions.cs]: This one worries me, since it's entirely likely the very first use of an ITextBuffer goes through this API (at least it is in my project system refactoring), which then means it's not going to get a workspace and never will. This may very well be causing a perf regression if the cloning stuff is necessary. Will probably have to ask for an RPS run then prior to this being merged unless we have an alternate solution, since taking an RPS regression now would cause codeflow to miss a few deadlines.

To be honest, it looks like this is all so we can call ITextBufferCloneService.Clone(ITextImage), which we could just have a static function that's Function<ITextImage, ITextBuffer> and we set that in VisualStudioWorkspace so it's OK for VS, and not set it in tests. Yes, statics are evil, but hey, at least instead of the static for PrimaryWorkspace now it's a single-use static that doesn't open pandora's box. Wishing I did that years ago now...

And that lives long enough until we pester @olegtk or somebody to give us a proper API here, because ITextImage.WithChanges() that returns an ITextImage would be both straightforward and also really useful.

Comment (sharwell) [reply src/EditorFeatures/Text/Extensions.cs]: Can you review commit c7eb42d1bc224106b4e563bb598a4e575f117d86 and let me know if the approach in that change is preferable? If so I will add it to this pull request.

File: src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs [hunk 19]
```diff
@@ -98,18 +111,11 @@ public override Encoding Encoding
...
                 {
-                    // simplest way to get text factory
-                    var ws = PrimaryWorkspace.Workspace;
-                    if (ws != null)
-                    {
-                        return ws.Services.GetService<ITextBufferCloneService>();
-                    }
-
-                    return null;
+                    return _workspace?.Services.GetService<ITextBufferCloneService>();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs]: i woudl feel better if this service was simply passed in in the start, rather than having to hold onto the _workspace itself.  is that doable?

Comment (sharwell) [reply src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs]: ➡️ I replied to Jason's post regarding this with a proposed alternative

**Comment by jaredpar** [comment]
Sorry my review status is flipping. Was experimenting with CodeFlow features.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs (hunk 3 above)
Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: Doesn't `LazyCancellation` only have an effect on continuations that specify a `CancellationToken`?

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.Connections.cs (hunk 8 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.Connections.cs]: I have no idea. I'm not familiar with this code and was simply trying to preserve behavior in the absence of static state class `PrimaryWorkspace`.

File: src/Features/Core/Portable/Shared/TestHooks/Legacy/ListenerForwarders.cs (hunk 1 above)
Comment (sharwell) [reply src/Features/Core/Portable/Shared/TestHooks/Legacy/ListenerForwarders.cs]: This change was intended to avoid the need to have **WorkspacesTestUtilities** reference `Features`. This is no longer possible, and **WorkspacesTestUtilities** ended up taking a reference all the way to **RemoteWorkspaces**. However, this change may still be beneficial as part of work to ensure all asynchronous operations started by Roslyn are tracked.

File: src/Workspaces/Core/Desktop/Workspace/Host/Mef/MefV1HostServices.cs (hunk 15 above)
Comment (sharwell) [reply src/Workspaces/Core/Desktop/Workspace/Host/Mef/MefV1HostServices.cs]: ➡️ Added comments in 37a1d2c

File: src/Workspaces/Core/Portable/Workspace/Host/Mef/MefHostServices.cs (hunk 16 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Workspace/Host/Mef/MefHostServices.cs]: ➡️ Added comments in 37a1d2c

File: src/Workspaces/Core/Portable/Workspace/Host/Mef/MefHostServices.cs (hunk 17 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Workspace/Host/Mef/MefHostServices.cs]: ➡️ Added comments in 37a1d2c

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs (hunk 12 above)
Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: These operations are only executed after the `_pendingWork` processing method completes. There isn't a way for it to be accessed at the same time.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: I don't think so (but i'm not totally sure). I thought it was more about the *antecedent* tasks and if *they* used a cancellation token.  in other words, if i have tasks:

```A <- B <- C```   (i.e. B depends on A and C depends on B)

And 'A' is running, but B gets canceled.  The question is: should 'C' run?  LazyCancellation (to me) says: "No.  All antecedent tasks are not completed.   Even though B transitioned ot Canceled, C should not run yet because 'A' is still running".  

This is regardless of the cancellation token passed to C.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: @stephentoub For guidance.

Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: 💭 I looked at referencesource.microsoft.com and it *seems* to only impact cases where a cancellation token is involved.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs (hunk 12 above)
Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: Fixed in 9c28f4a

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs (hunk 11 above)
Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: :arrow_right: Fixed in 9c28f4a

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs (hunk 10 above)
Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: :arrow_right: Fixed in 9c28f4a

File: src/Workspaces/Core/Portable/Workspace/PrimaryWorkspace.cs (hunk 7 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Workspace/PrimaryWorkspace.cs]: :arrow_right: Fixed in 9c28f4a

File: src/Workspaces/Core/Portable/Workspace/PrimaryWorkspace.cs (hunk 6 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Workspace/PrimaryWorkspace.cs]: :arrow_right: Fixed in 9c28f4a

File: src/EditorFeatures/Text/Extensions.cs (hunk 18 above)
Comment (sharwell) [reply src/EditorFeatures/Text/Extensions.cs]: ➡️ Fixed in 4924d70

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs (hunk 3 above)
Comment (stephentoub) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: @CyrusNajmabadi is correct re "I thought it was more about the antecedent tasks and if they used a cancellation token." If you have:
```C#
Task a = ...;
Task b = a.ContinueWith(..., token);
Task c = b.ContinueWith(...);
```
and while a is running the token has cancellation requested, should c be allowed to run immediately or must it wait for a to finish.

Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: @stephentoub I believe you may have misunderstood the context. In my code, I pass `CancellationToken.None` to `ContinueWith`. I am claiming that `TaskContinuationOptions.LazyCancellation` will never have an impact on a call to `ContinueWith` called in this manner.

Comment (stephentoub) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: I didn't look at your code; I was commenting on the behavior of LazyCancellation and what it actually does.

Given the code:
```C#
Task a = ...;
Task b = a.ContinueWith(..., token);
Task c = b.ContinueWith(...);
```
`c` can end up running concurrently with `a`.  To prevent that, you can make it:
```C#
Task a = ...;
Task b = a.ContinueWith(..., token, TaskContinuationOptions.LazyCancellation, TaskScheduler.Default);
Task c = b.ContinueWith(...);
```
and then `c` will not run concurrently with `a`.  If that `token` is `None`, then it doesn't matter because it won't be canceled prior to `a` completing (it could still be canceled by other means, e.g. if you passed TaskContinuationOptions.OnlyOnFaulted and the antecedent completed in another state, but in such cases the antecedent needs to be completed for that cancellation to happen).

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.SemanticChangeProcessor.cs]: If this is the case, can we comment that it's ok for us to not pass LazyCancellation here as it seems to only be relevant for the ContinueWith's that take cancellation tokens?

Thanks!

File: src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs [hunk 20]
```diff
@@ -213,7 +203,7 @@ public override SourceText WithChanges(IEnumerable<TextChange> changes)
                 }
 
                 // check whether we can use text buffer factory
-                var factory = TextBufferFactory;
+                var factory = _textBufferCloneServiceOpt;
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs]: Nit: now that that's readonly, we don't have to worry about capturing to a local which was previously required. (But hey, it's shorter than inlining, so...)

File: src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs [hunk 21]
```diff
@@ -82,8 +80,8 @@ public static SourceText From(Workspace workspace, ITextSnapshot editorSnapshot)
                 if (!s_textSnapshotMap.TryGetValue(editorSnapshot, out var snapshot))
                 {
                     // Avoid capturing `workspace` on the fast path
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs]: Nit: comment name needs updating.

File: src/EditorFeatures/Text/Extensions.cs [hunk 22]
```diff
@@ -37,10 +37,10 @@ internal static TextLine AsTextLine(this ITextSnapshotLine line)
             => line.Snapshot.AsText().Lines[line.LineNumber];
 
         public static SourceText AsText(this ITextSnapshot textSnapshot)
-            => SnapshotSourceText.From(workspace: null, textSnapshot);
+            => SnapshotSourceText.From(textBufferCloneServiceOpt: null, textSnapshot);
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Text/Extensions.cs]: Should this try fetching the service like TextBufferContainer does?

File: src/EditorFeatures/Text/Implementation/TextBufferFactoryService/TextBufferCloneServiceFactory.cs [hunk 23]
```diff
@@ -16,25 +16,27 @@ internal class TextBufferCloneServiceFactory : IWorkspaceServiceFactory
...
-            IContentTypeRegistryService contentTypeRegistry)
+            IContentTypeRegistryService contentTypeRegistryService)
         {
-            _singleton = new TextBufferCloneService((ITextBufferFactoryService3)textBufferFactoryService, contentTypeRegistry.UnknownContentType);
+            _singleton = new TextBufferCloneService((ITextBufferFactoryService3)textBufferFactoryService, contentTypeRegistryService);
         }
 
         public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
         {
             return _singleton;
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Text/Implementation/TextBufferFactoryService/TextBufferCloneServiceFactory.cs]: _singleton seems a misnomer here: there's one you'll get through the factory, and one directly from the export below. This could just have the [ImportingConstructor] import the type below and return it, just making this an even thinner shim.

(If the only thing still using the workspace service is that inline rename thing, then just disregard this comment and we'll delete this then.)

File: src/EditorFeatures/Text/Extensions.cs (hunk 22 above)
Comment (sharwell) [reply src/EditorFeatures/Text/Extensions.cs]: Unfortunately `ITextSnapshot` doesn't expose any properties to try and fetch from.

File: src/EditorFeatures/Text/Implementation/TextBufferFactoryService/TextBufferCloneServiceFactory.cs (hunk 23 above)
Comment (sharwell) [reply src/EditorFeatures/Text/Implementation/TextBufferFactoryService/TextBufferCloneServiceFactory.cs]: ➡️ Covered as part of #25659

File: src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs (hunk 21 above)
Comment (sharwell) [reply src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs]: I'll update this in a commit in my follow-up pull request to avoid needing to rebuild everything again.

File: src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs (hunk 20 above)
Comment (sharwell) [reply src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs]: I would likely change it (at least rename) if the code was more complex, but the method seems straightforward enough to keep.

**Comment by sharwell** [comment]
@Pilchie and/or @MattGertz for approval

**Comment by Pilchie** [comment]
Have we verified that TypeScript/F#/Xaml work with these changes?  Do they have an impact on VS4Mac that we should worry about?

**Comment by sharwell** [comment]
@Pilchie Not specifically. The listener-related items valuable in testing scenarios have type forwarders and shouldn't be a problem unless Workspaces needs some other IVT that Features already has. The other internal API changes were primarily used in remote workspace / OOP scenarios so it's unlikely to cause problems.

**Comment by Pilchie** [comment]
Tagging @minestarks @brettfo @mgoertz-msft - can you three take a look and chime in on whether you think this will impact your scenarios at insertion time?

**Comment by mgoertz-msft** [comment]
@Pilchie Shouldn't affect the XAML language service as long as you don't change any of the interfaces in XamlVisualStudio.csproj. The only other concern would be if this were to affect CustomWorkspaces, which we do use in the VS codebase.

**Comment by brettfo** [comment]
Looking through the changes I don't see anything that'll likely break F#, but if I could be emailed when we have a VAL build I'll run through some scenarios.

**Comment by Pilchie** [comment]
Approved - thanks for the diligence.
